orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -843,3 +854,306 @@ def _update_onnx_model_initializers(self, state_tensors):
+
+        model = 'model'
+        model_states = self._training_session.get_model_state(include_mixed_precision_weights=False)
+        states[model] = {}
+
+        # extract trainer model weights from the training session
+        for precision in model_states:
+            states[model][precision] = {}
+            for model_state_key in model_states[precision]:
+                states[model][precision][model_state_key] = torch.from_numpy(model_states[precision][model_state_key])
+
+        # extract untrained (frozen) model weights
+        precision_map = {TensorProto.FLOAT: 'fp32'}
+        for node in self._onnx_model.graph.initializer:
+            if node.name not in states[model][precision] and node.name in self.options.utils.frozen_weights:
+                if node.data_type in precision_map:
+                    states[model][precision_map[node.data_type]][node.name] = \
+                        torch.from_numpy(np.array(onnx.numpy_helper.to_array(node)))
+
+    def _extract_optimizer_states(self, states):
+        """Extract optimizer states from the training session and load into the states dictionary"""
+
+        optimizer = 'optimizer'
+        optimizer_states = self._training_session.get_optimizer_state()
+        states[optimizer] = {}
+        for model_state_key in optimizer_states:
+            states[optimizer][model_state_key] = {}
+            for optimizer_state_key in optimizer_states[model_state_key]:
+                states[optimizer][model_state_key][optimizer_state_key] = \
+                    torch.from_numpy(optimizer_states[model_state_key][optimizer_state_key])
+
+    def _extract_trainer_options(self, states):
+        """Extract relevant trainer configuration and load it into the states dictionary"""
+
+        trainer_options = 'trainer_options'
+        states[trainer_options] = {}
+        states[trainer_options]['mixed_precision'] = self.options.mixed_precision.enabled
+        states[trainer_options]['zero_stage'] = self.options.distributed.deepspeed_zero_optimization.stage or 0
+        states[trainer_options]['world_rank'] = self.options.distributed.world_rank or 0
+        states[trainer_options]['world_size'] = self.options.distributed.world_size or 1
+        states[trainer_options]['optimizer_name'] = self.optim_config.name
+
+    def state_dict(self, pytorch_format=False):
+        """Extracts model and optimizer states from the training session and returns them in a dictionary
+
+        - Extracts model and optimizer states from the training session
+        - Extracts relevant trainer options
+        - Extracts partition information from the training session in case of a ZeRO enabled distributed
+            ORTTrainer configuration
+        - Loads them into a dictionary and returns this dictionary
+
+        Args:
+            pytorch_format: boolean flag to indicate the format of the state dictionary to return.
```

Comment:
```suggestion
        """Returns a dictionary with model, and optionally, optimizer states

        The returned dictionary contains the following information:
        - Model and optimizer states
        - Required ORTTrainerOptions settings
        - Distributed training information, such as but not limited to ZeRO

        Args:
            pytorch_format: boolean flag to select either ONNX Runtime or PyTorch state schema
```

Try avoid mentioning internal information, such as training session or use verbs such as "extract" that gives a hint on how the operation is done. Developers, specially python, leverages this kind of information to hack things :)